fw/http_frame.c

```diff
@@ -641,10 +646,20 @@ tfw_h2_conn_streams_cleanup(TfwH2Ctx *ctx)
 
 	WARN_ON_ONCE(((TfwConn *)conn)->stream.msg);
 
+	T_DBG3("%s: ctx [%p] conn %p sched %p\n", __func__, ctx, conn, sched);
+
 	rbtree_postorder_for_each_entry_safe(cur, next, &sched->streams, node) {
 		tfw_h2_stream_unlink(ctx, cur);
-		tfw_h2_stream_clean(ctx, cur);
+
+		/* The streams tree is about to be destroyed and
+		 * we don't want to trigger rebalancing.
+		 * No further actions regarding streams dependencies/prio
+		 * is required at this stage.
+		 */
+		tfw_h2_delete_stream(cur);
+		--ctx->streams_num;
```

Comment:
I'm wondering why do we traverse the tree and delete all the streams one by one instead of using TfwPool for the connection? `TfwStream` looks not so small, so probably we can easily utilize a page for a connection. Are they any other allocations with the connection life time?

Can we avoid the loop at all and just do `ctx->streams_num = 0`?

Reply:
I believe it was just an initial design decision to use streams `kmem_cache` for all connections. I would create a separate task to evaluate performance.